release/release.go

```diff
@@ -0,0 +1,199 @@
+package main
+
+import (
+	"archive/tar"
+	"archive/zip"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/mongodb/mongo-tools/release/platform"
+)
+
+func main() {
+	if len(os.Args) != 2 {
+		fail("please provide exactly one subcommand name")
+	}
+	cmd := os.Args[1]
+
+	switch cmd {
+	case "build-archive":
+		buildTarball()
+		buildZip()
+	case "build-installer":
+		fail("not implemented")
+	case "build-linux":
+		fail("not implemented")
+	}
+}
+
+var binaries = []string{
+	"bsondump",
+	"mongodump",
+	"mongoexport",
+	"mongofiles",
+	"mongoimport",
+	"mongorestore",
+	"mongostat",
+	"mongotop",
+}
+
+var staticFiles = []string{
+	"LICENSE.md",
+	"README.md",
+	"THIRD-PARTY-NOTICES",
+}
+
+func log(format string, args ...interface{}) {
+	fmt.Fprintf(os.Stderr, format+"\n", args...)
+}
+
+func check(err error, format ...interface{}) {
+	if err == nil {
+		return
+	}
+	msg := err.Error()
+	if len(format) != 0 {
+		task := fmt.Sprintf(format[0].(string), format[1:]...)
+		msg = fmt.Sprintf("'%s' failed: %v", task, err)
+	}
+	fail(msg)
+}
+
+func fail(format string, args ...interface{}) {
+	log(format, args...)
+	os.Exit(1)
+}
+
+func run(name string, args ...string) (string, error) {
+	cmd := exec.Command(name, args...)
+	out, err := cmd.Output()
+	return strings.TrimSpace(string(out)), err
+}
+
+func getVersion() string {
+	desc, err := run("git", "describe")
+	check(err, "git describe")
+	return desc
+}
+
+func getReleaseName() string {
+	p, err := platform.Get()
+	check(err, "get platform")
+	version := getVersion()
+
+	return fmt.Sprintf(
+		"mongodb-cli-tools-%s-%s-%s",
+		p.Name, p.Arch, version,
+	)
+}
+
+func addToTarball(tw *tar.Writer, dst, src string) {
+	file, err := os.Open(src)
+	check(err, "open file")
+	defer file.Close()
+
+	stat, err := file.Stat()
+	check(err, "stat file")
+
+	header := &tar.Header{
+		Name: dst,
+		Size: stat.Size(),
+		Mode: int64(stat.Mode()),
+	}
+
+	err = tw.WriteHeader(header)
+	check(err, "write header to archive")
+
+	_, err = io.Copy(tw, file)
+	check(err, "write file to archive")
+}
+
+func buildTarball() {
+	win, err := platform.IsWindows()
+	check(err, "check platform type")
+	if win {
+		log("skipping tarball creation on windows")
+		return
+	}
+
+	releaseName := getReleaseName()
+	archiveFile, err := os.Create(releaseName + ".tgz")
+	check(err, "create archive file")
+	defer archiveFile.Close()
+
+	gw := gzip.NewWriter(archiveFile)
+	defer gw.Close()
+
+	tw := tar.NewWriter(gw)
+	defer tw.Close()
+
+	for _, name := range staticFiles {
+		log("adding %s to tarball", name)
+		src := name
+		dst := filepath.Join(releaseName, name)
+		addToTarball(tw, dst, src)
+	}
+
+	for _, binName := range binaries {
+		log("adding %s binary to tarball", binName)
+		src := filepath.Join(".", "bin", binName)
```

Comment:
minor: here you use "." and above in line 137, you don't.   Also, I might have abstracted "write a list of files with an optional prefix" as the two for loops are otherwise identical.  E.g.:
```
addFilesToTarball("", staticFiles)
addFilesToTarball("bin",binaries)
```
But I'm not sure it's worth rewriting.

Reply:
removed the `.` for consistency; I thought about factoring it further, but decided to keep the duplication since the loops were so small

Reply:
removed the `.` for consistency. thought about that rewrite, but decided to leave the duplication instead of abstracting since the loops were so small